README.md
# EvermoreDev's React Boilerplate
[![Build Status](https://travis-ci.org/evermoredev/react-boilerplate.svg?branch=master)](https://travis-ci.org/evermoredev/react-boilerplate) 

This React boilerplate is designed to get you up and running with all of the latest technologies in the React world.

## Technologies include:

* [React 15.3](https://github.com/facebook/react)
* [React Router 4](https://react-router.now.sh)
* [Babel 6](https://babeljs.io)
* [Jest](https://facebook.github.io/jest/docs/getting-started.html)
* [PostCSS-Loader](https://github.com/postcss/postcss-loader)
* [Webpack](https://webpack.github.io)
* [Webpack Dev Server](https://webpack.github.io/docs/webpack-dev-server.html)
* [React Transform](https://github.com/gaearon/react-transform-hmr)

## Usage

```
npm install
npm start
http://localhost:4000
```

## Features

### [React](https://github.com/facebook/react)

The newest version of React with ES2016 modules and classes make your React components clean and easy to follow.

```jsx
import React from 'react';

class HomeView extends React.Component {

  render() {
    return (
      <div className="home-view">
        <h1>Welcome Home!</h1>
      </div>
    );
  }

}

export default HomeView;
```

### [React Router 4](https://react-router.now.sh)

Two example routes have already been built in, as well as a 404/Not Found page.

### [Babel 6](http://babeljs.io)

Already configured to use the latest ES2016 and ES2017 features.

### [Jest](https://facebook.github.io/jest/docs/getting-started.html)

Uses React's testing library [Jest](https://facebook.github.io/jest/docs/getting-started.html). Tests are set up out of the box. Simply run `npm test` to see Jest in action. Initial tests are located in `__tests__/` and follow the same folder structure as `src/`.

The tests also provide a simple example of using airbnb's Testing Utility [Enzyme](http://airbnb.io/enzyme/index.html) to mount React components.

### [PostCSS-Loader](https://github.com/postcss/postcss-loader)

Styling is set up for [Sass](http://sass-lang.com/) in the `src/components/styles` directory and the building is already configured. We've included [normalize.css](https://github.com/necolas/normalize.css/) for a good reset.

### [Webpack Dev Server](http://webpack.github.io/docs/webpack-dev-server.html)

From the docs:

> The webpack-dev-server is a little Node.js Express server, which uses the webpack-dev-middleware to serve a webpack bundle. It also has a little runtime which is connected to the server via Sock.js.

## Still More...

### ESLint

Yeah, we're a little opinionated. We've added a lot of rules and always looking to add more. Make some suggestions if you have them!

### CHANGELOG.md

We've included a CHANGELOG with the project that we're keeping up with but highly recommend its use for your own project. Clean it out and start your own!

### [Travis](https://docs.travis-ci.com/)

We've included continuous integration testing with [Travis](https://docs.travis-ci.com/). Be sure to enable it in your repo.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased] (Work toward 1.1.0)
### Added
- static folder with a `README.md` about its use

### Changed

### Fixed

### Removed

### Known Issues 

## [1.0.0] - 2017-01-04

### Added
- First commit of project


__tests__/components/views/HomeView.test.js
import React from 'react';
import { mount } from 'enzyme';

import { HomeView } from '../../../src/components/views';

describe('HomeView Component', () => {

  let wrapper;

  // Mount the component before each test in case we modify it
  beforeEach(() => {
    wrapper = mount(
      <HomeView />
    );
  });

  it('Renders "Welcome Home!" text', () => {
    expect(wrapper.find('h1').text()).toEqual('Welcome Home!');
  });

});


__tests__/components/views/AboutView.test.js
import React from 'react';
import { mount } from 'enzyme';

import { AboutView } from '../../../src/components/views';

describe('AboutView Component', () => {

  let wrapper;

  // Mount the component before each test in case we modify it
  beforeEach(() => {
    wrapper = mount(
      <AboutView />
    );
  });

  it('Renders "About Us!" text', () => {
    expect(wrapper.find('h1').text()).toEqual('About Us!');
  });

});


__tests__/components/views/NotFoundView.test.js
import React from 'react';
import { mount } from 'enzyme';

import { NotFoundView } from '../../../src/components/views';

describe('NotFoundView Component', () => {

  let wrapper;

  // Mount the component before each test in case we modify it
  beforeEach(() => {
    wrapper = mount(
      <NotFoundView />
    );
  });

  it('Renders "404: Not Found!" text', () => {
    expect(wrapper.find('h1').text()).toEqual('404: Not Found!');
  });

});


src/App.js
import React from 'react';
import { BrowserRouter, Match, Miss, Link } from 'react-router';

// Import view components
import {
  ChartView,
  HomeView,
  NotFoundView,
  ResultsView
} from './components/views/index.js';

export default class App extends React.Component {

  render() {
    return (
      <BrowserRouter>
        <div className="wrapper">
          <nav className="main-nav">
            <ul>
              <li>PokerApp</li>
            </ul>
          </nav>

          <Match exactly pattern="/" component={HomeView} />
          <Match pattern="/chart" component={ChartView} />
          <Match pattern="/results" component={ResultsView} />

          {/* If none of those match, then a sibling `Miss` will render. */}
          <Miss component={NotFoundView} />
          <footer>
            <ul>
              <li>
                <Link to="/">
                  <i className="fa fa-home" aria-hidden="true" />
                  <span>Home</span>
                </Link>
              </li>
              <li>
                <Link to="/results">
                  <i className="fa fa-area-chart" aria-hidden="true" />
                  <span>Results</span>
                </Link>
              </li>
              <li>
                <Link to="/chart">
                  <i className="fa fa-bar-chart" aria-hidden="true" />
                  <span>Charts</span>
                </Link>
              </li>
              <li>
                <a href="#">
                  <i className="fa fa-bars" aria-hidden="true" />
                  <span>More</span>
                </a>
              </li>
            </ul>
          </footer>
        </div>
      </BrowserRouter>
    );
  }

}


webpack.config.production.js
/**
 * WEBPACK CONFIG
 *
 * Notes on config properties:
 *
 * 'entry'
 * Entry point for the bundle.
 *
 * 'output'
 * If you pass an array - the modules are loaded on startup. The last one is exported.
 *
 * 'resolve'
 * Array of file extensions used to resolve modules.
 *
 * devtool: 'eval-source-map'
 * http://www.cnblogs.com/Answer1215/p/4312265.html
 * The source map file will only be downloaded if you have source maps enabled
 * and your dev tools open.
 *
 * OccurrenceOrderPlugin
 * Assign the module and chunk ids by occurrence count. Ids that are used often
 * get lower (shorter) ids. This make ids predictable, reduces to total file size
 * and is recommended.
 *
 * UglifyJsPlugin
 * Minimize all JavaScript output of chunks. Loaders are switched into minimizing mode.
 *    - 'compress'
 *      Compressor is a tree transformer which reduces the code size by applying
 *      various optimizations on the AST.
 *
 * 'NODE_ENV'
 * React relies on process.env.NODE_ENV based optimizations.
 * If we force it to production, React will get in an optimized manner.
 * This will disable some checks (eg. property type checks) and give you a smaller
 *    build and improved performance.
 *
 *    Note: That JSON.stringify is needed as webpack will perform string replace
 *    "as is". In this case we'll want to end up with strings as that's what
 *    various comparisons expect, not just production. Latter would just cause
 *    an error.
 *
 * 'babel'
 * Babel enables the use of ES6 today by transpiling your ES6 JavaScript into
 * equivalent ES5 source that is actually delivered to the end user browser.
 */

const webpack = require('webpack');
const path = require('path');

module.exports = {
  entry: './src/index',
  output: {
    path: path.join(__dirname, 'dist'),
    filename: 'bundle.js',
    publicPath: '/'
  },
  resolve: {
    extensions: ['', '.js', '.jsx']
  },
  devtool: 'source-map',
  plugins: [
    new webpack.optimize.OccurenceOrderPlugin(),
    new webpack.DefinePlugin({
      'process.env': {
        NODE_ENV: JSON.stringify('production')
      }
    }),
    new webpack.optimize.UglifyJsPlugin({
      compress: {
        warnings: false
      }
    })
  ],
  module: {
    loaders: [
      {
        test: /\.jsx?$/,
        loaders: ['babel'],
        include: path.join(__dirname, 'src')
      },
      {
        test: /\.[s]css$/,
        loaders: [
          'style-loader',
          'css-loader?importLoaders=1',
          'postcss-loader?sourceMap=inline'
        ]
      }
    ]
  },
  postcss: function (webpack) {
    return [
      require('postcss-smart-import')({
        addDependencyTo: webpack
      }),
      require('precss')(),
      require('autoprefixer')(),
    ];
  }
};


webpack.config.js
/**
 * WEBPACK CONFIG
 *
 * Notes on config properties:
 *
 * 'entry'
 * Entry point for the bundle.
 *
 * 'output'
 * If you pass an array - the modules are loaded on startup. The last one is exported.
 *
 * 'resolve'
 * Array of file extensions used to resolve modules.
 *
 * 'webpack-dev-server'
 * Is a little node.js Express server, which uses the webpack-dev-middleware to
 * serve a webpack bundle. It also has a little runtime which is connected to
 * the server via Socket.IO.
 *
 * 'webpack/hot/dev-server'
 * By adding a script to your index.html file and a special entry point in your
 * configuration you will be able to get live reloads when doing changes to your
 * files.
 *
 * devtool: 'eval-source-map'
 * http://www.cnblogs.com/Answer1215/p/4312265.html
 * The source map file will only be downloaded if you have source maps enabled
 * and your dev tools open.
 *
 * HotModuleReplacementPlugin()
 * Hot Module Replacement (HMR) exchanges, adds or removes modules while an
 * application is running without page reload.
 *
 * NoErrorsPlugin()
 * Hot loader is better when used with NoErrorsPlugin and hot/only-dev-server
 * since it eliminates page reloads altogether and recovers after syntax errors.
 *
 * 'react-hot'
 * React Hot Loader is a plugin for Webpack that allows instantaneous live
 * refresh without losing state while editing React components.
 *
 * 'babel'
 * Babel enables the use of ES6 today by transpiling your ES6 JavaScript into equivalent ES5 source
 * that is actually delivered to the end user browser.
 */

const webpack = require('webpack');
const path = require('path');

module.exports = {
  entry: [
    'webpack-dev-server/client?http://localhost:4000',
    'webpack/hot/dev-server',
    './src/index'
  ],
  output: {
    path: __dirname,
    filename: 'bundle.js',
    publicPath: '/static/'
  },
  resolve: {
    extensions: ['', '.js', '.jsx']
  },
  devtool: 'eval-source-map',
  plugins: [
    new webpack.HotModuleReplacementPlugin(),
    new webpack.NoErrorsPlugin()
  ],
  module: {
    loaders: [
      {
        test: /\.jsx?$/,
        loaders: ['babel'],
        include: path.join(__dirname, 'src')
      },
      {
        test: /\.[s]css$/,
        loaders: [
          'style-loader',
          'css-loader?importLoaders=1',
          'postcss-loader?sourceMap=inline'
        ]
      }
    ]
  },
  postcss: function (webpack) {
    return [
      require('postcss-smart-import')({
        addDependencyTo: webpack
      }),
      require('precss')(),
      require('autoprefixer')(),
    ];
  }
};
